client/hotrod-client/src/main/java/org/infinispan/client/hotrod/impl/transport/netty/ChannelFactory.java

```diff
@@ -110,23 +112,30 @@ public class ChannelFactory {
    private CompletableFuture<Void> clusterSwitchStage;
    // Servers for which the last connection attempt failed and which have no established connections
    @GuardedBy("lock")
-   private final Set<SocketAddress> failedServers = new HashSet<>();
+   private final Set<SocketAddress> failedServers;
+   private final Configuration configuration;
    private final CodecHolder codecHolder;
    private AddressResolverGroup<?> dnsResolver;
    private SslContext sslContext;
    private FileWatcher watcher;
 
-   public ChannelFactory(CodecHolder codecHolder) {
+   public ChannelFactory(Configuration configuration, CodecHolder codecHolder) {
+      this.configuration = configuration;
       this.codecHolder = codecHolder;
+
+      this.failedServers = configuration.clientIntelligence() == ClientIntelligence.BASIC ?
```

Comment:
The problem is not limited to BASiC

Reply:
I will take a look next week, thanks.

Reply:
@wburns have you had a chance to try reproducing the problem with non basic ? you will need to use a server build without https://github.com/infinispan/infinispan/pull/12912

Reply:
Sorry, I was able to finally look into this. I added a new test method and it is behaving as I would expect.

```java
   public void testHashAware() {
      org.infinispan.client.hotrod.configuration.ConfigurationBuilder clientBuilder =
            new org.infinispan.client.hotrod.configuration.ConfigurationBuilder();
      clientBuilder.addServers(HotRodClientTestingUtil.getServersString(server(0), server(1)));
      clientBuilder.clientIntelligence(ClientIntelligence.HASH_DISTRIBUTION_AWARE);

      withRemoteCacheManager(new RemoteCacheManagerCallable(
            new InternalRemoteCacheManager(clientBuilder.build())) {
         @Override
         public void call() {
            RemoteCache<Object, Object> cache = rcm.getCache();
            ChannelFactory cf = rcm.getChannelFactory();
            assertFalse(cache.containsKey("k"));
            AtomicBoolean shouldEnd = new AtomicBoolean();
            Future<?> f = testExecutor().submit(() -> {
               try {
                  while (!shouldEnd.get()) {
                     HotRodServer server = servers.get(1);
                     log.fatal("Killing server");
                     killServer(1);
                     addHotRodServer(builder, server.getPort());
                     log.fatal("New server added");

                     Thread.sleep(2_000);
                  }
               } catch (InterruptedException e) {
                  throw new RuntimeException(e);
               }
            });
            for (int i = 0; i < 100_000; i++) {
               assertFalse(cache.containsKey("k" + i));
            }
            eventuallyEquals(0, () -> cf.getFailedServers().size());

            eventuallyEquals(0, () -> {
               assertFalse(cache.containsKey("k"));
               return cf.getFailedServers().size();
            });

            shouldEnd.set(true);
            try {
               f.get(10, TimeUnit.SECONDS);
            } catch (InterruptedException | ExecutionException | TimeoutException e) {
               throw new RuntimeException(e);
            }
         }
      });
   }
   ```
   
After reverting the commit that fixes the server, I am able to see the spam of the suspected messages. But this only happens only for a very short duration (only seen because I am forcing my code to add delays - 1 sec to install view). It resolves itself as soon as the server publishes itself into the topology view. This allows the server to send the topology information back to the client so it can update.

The trace can be seen here https://gist.github.com/wburns/d1326b770c78baa3f6fee8a4bc6ffaa9 Note the rest of the test runs just fine. Note this test kills and readds the server over and over and queries different keys so it will hit them all.

The problem is until a client first retrieves the topology info it basically has to run in mode akin to BASIC so it has the same issue, albeit temporary, this PR fixes with BASIC.

Reply:
> It resolves itself as soon as the server publishes itself into the topology view. 

In your test, you are killing and starting a server in a loop, forcing clients to constantly update the topology. The topology update serves the same purpose as the fix in this PR – forcing a new connection attempt. In my production environment, topology changes happen rarely, so a failed server remains in the failedServers set for hours or even days.

Reply:
I changed the test to only kill the server once and it still resolved properly as well every time I ran it.

Reply:
I was able to reproduce a problem with not recovering failed server. To force topology update before server is ready i modified HotRodServer.java. With real server stating manually I do not need any code modifications to get a connection refused during topology update in client, but i does not happen in test environment.

```
// Add self to topology cache last, after everything is initialized
      if (Configurations.isClustered(SecurityActions.getCacheManagerConfiguration(cacheManager))) {
         defineTopologyCacheConfig(cacheManager);
         if (log.isDebugEnabled())
            log.debugf("Externally facing address is %s:%d", configuration.proxyHost(), configuration.proxyPort());

         addSelfToTopologyView(cacheManager);
      }

       try {
           Thread.sleep(50);
       } catch (InterruptedException e) {
           throw new RuntimeException(e);
       }

      super.startInternal();
```

here is my test : 

```
public void testHashAware() {
      org.infinispan.client.hotrod.configuration.ConfigurationBuilder clientBuilder =
              new org.infinispan.client.hotrod.configuration.ConfigurationBuilder();
      clientBuilder.addServers(HotRodClientTestingUtil.getServersString(server(0), server(1)));
      clientBuilder.clientIntelligence(ClientIntelligence.HASH_DISTRIBUTION_AWARE);

      withRemoteCacheManager(new RemoteCacheManagerCallable(
              new InternalRemoteCacheManager(clientBuilder.build())) {
         @Override
         public void call() {
            RemoteCache<Object, Object> cache = rcm.getCache();
            ChannelFactory cf = rcm.getChannelFactory();
            assertFalse(cache.containsKey("k"));
            AtomicBoolean shouldEnd = new AtomicBoolean();
            testExecutor().execute(() -> {
               try {
                  Thread.sleep(2_000);
                  HotRodServer server = servers.get(1);
                  log.fatal("Killing server");
                  killServer(1);
                  Thread.sleep(500);
                  log.fatal("Adding server");
                  addHotRodServer(getCacheConfiguration(), server.getPort());
                  Thread.sleep(2_000);
                  shouldEnd.set(true);
               } catch (InterruptedException e) {
                  throw new RuntimeException(e);
               }
            });

            for (int i = 0; i < 100_000 || shouldEnd.get(); i++) {
               assertFalse(cache.containsKey("k" + i));
               if (!cf.getFailedServers().isEmpty()) {
                  log.fatal("failedServers=" + cf.getFailedServers().size() + " " + i);
               }
               try {
                  Thread.sleep(10);
               } catch (InterruptedException e) {
                  throw new RuntimeException(e);
               }
            }
         }
      });
```

log:

```
15:12:14,657 FATAL (ForkThread-1,ConsistentHashTest) [o.i.t.AbstractInfinispanTest] Killing server
15:12:15,412 FATAL (ForkThread-1,ConsistentHashTest) [o.i.t.AbstractInfinispanTest] Adding server
15:12:15,621 FATAL (testng-ConsistentHashTest -- key-type=text/plain) [o.i.t.AbstractInfinispanTest] failedServers=1 201
15:12:15,635 FATAL (testng-ConsistentHashTest -- key-type=text/plain) [o.i.t.AbstractInfinispanTest] failedServers=1 202
15:12:15,648 FATAL (testng-ConsistentHashTest -- key-type=text/plain) [o.i.t.AbstractInfinispanTest] failedServers=1 203
15:12:15,663 FATAL (testng-ConsistentHashTest -- key-type=text/plain) [o.i.t.AbstractInfinispanTest] failedServers=1 204
```
`failedServers=1` logged until the test ends.

Reply:
> I changed the test to only kill the server once and it still resolved properly as well every time I ran it.
In your test, you get a failedServer when you kill the server, and the topology update triggered by this event removes the server from the topology and the failedServer set. You cannot see the failedServer update caused by the addServer call due to limitations of the test framework.

Reply:
Thanks, I will take a look.

Reply:
Your test is not testing the client, this is testing/verifying ISPN-12828, which is already fixed.

When using a REHASH_AWARE/TOPOLOGY_AWARE client we support adding/removing servers based on the responses of the servers. Since this is caused by the server misreporting what is actually running, this is not an issue with the client itself.

The fix in this PR is for BASIC intelligence solely, since it has no reliable way to remedy a single server coming or going (requires multiple configured). Honestly, I debate if this PR is even needed for BASIC as most use cases of BASIC use a single server host (usually a load balancer), which doesn't have the problem this PR fixes.

Reply:
@wburns I am testing the behavior of the client when it is unable to connect to a new server added to the topology. This could occur due to a server bug or a random network hiccup. The client should be able to recover from this.

Reply:
The client can recover in many cases including various server bugs and network hiccups. When either of those happen they will almost certainly cause another topology update to be sent back to the client.

But I was thinking about it this evening and I don't think there would be any cause in making this option apply to all intelligences, but I will have to take a look again. And would make the option less confusing, which is more what I like tbh.

Reply:
> When either of those happen they will almost certainly cause another topology update to be sent back to the client.

My test shows an example of unrecoverable hiccup.

Reply:
I think we have different definitions of unrecoverable in this case, but I have changed it to be for all intelligence types as I think it will also be less confusing from a documentation and code perspective.